Added WMS Button functionality

tony-ross opened the pull request.

tony-ross committed:
Added WMS Button functionality

h-m-jones commented on grails-app/views/occurrence/_map.gsp:
```diff
@@ -391,6 +430,29 @@
+        $('#copyWmsParameters').click(function() {
+            var wmsParams = $('#wmsParams');
+            wmsParams.select();
+            document.execCommand('copy');
```
document.execCommand is now deprecated and you have to use navigator.clipboard instead which is async, so I asked HAL and he replaced with:

$('#copyWmsParameters').click(function () {
    const wmsParams = $('#wmsParams').val(); // Get the value of the textarea
    const button = $(this);
    const originalText = button.text();

    // Use Clipboard API to copy text
    navigator.clipboard.writeText(wmsParams)
        .then(() => {
            // Show temporary success message
            button.text('<g:message code="map.wms.btn.copied" default="Copied!"/>');
            setTimeout(() => {
                button.text(originalText);
            }, 2000);
        })
        .catch(err => {
            console.error('Failed to copy text: ', err);
            alert('Failed to copy text. Please try again.');
        });
});

tony-ross replied:
Done

h-m-jones commented on grails-app/views/occurrence/_map.gsp:
```diff
@@ -1145,6 +1207,79 @@
+            var baseUrl = MAP_VAR.mappingUrl + "/mapping/wms/reflect";
+
+            // Format WMS parameters in tabular format
+            var formattedParams = Object.keys(wmsParams).map(function(key) {
```
You can chain map and filter, so something like:

const formattedParams = Object.keys(wmsParams)
    .map(key => (key.toUpperCase() + ':').padEnd(20, ' ') + wmsParams[key])
    .filter(line =>
        !line.startsWith('SIZE:') &&
        !line.startsWith('STYLE:') &&
        !line.startsWith('OUTLINE:') &&
        !line.startsWith('COLOUR:')
    )
    .sort()
    .join('\n');

tony-ross replied:
Done

h-m-jones commented on grails-app/views/occurrence/_map.gsp:
```diff
@@ -1145,6 +1207,79 @@
+        }
+    }
+
+    // Add event listeners to map controls that trigger WMS updates
```
I've no idea, but my first thought was why do you need these listeners because can't you just update the modal content when it opens?

tony-ross replied:
Your right the listeners are not required. I'll remove them.

h-m-jones commented on grails-app/views/occurrence/_map.gsp:
A couple of comments, but nice work. Very clean. This will need to be deployed to the mini-atlas and tested before merging to pipelines. We'll sort this out

tony-ross replied:
I have moved the WMS Button functionality out of the _map.gsp and into a new Tag.

tony-ross committed:
removed event listners
replaced deprecated document.execCommand
used chained map and filter expressions

tony-ross committed:
correct query parameter

tony-ross committed:
Modularise WMS Button functionality
- Created a new WMS button Grails tag. Used in the map view that triggers a modal for generating WMS query URLs etc.
- Moved WMS modal HTML to a separate file - better organisation and maintainability.
- Re-Implemented JavaScript functions to handle WMS parameter updates and clipboard copying.
- Removed redundant WMS modal code

tony-ross committed:
tidy up - remove lines

tony-ross committed:
tidy up - remove lines

tony-ross committed:
Added map tool changes to ENV
Updated clipboard copying method to use modern API with fallback for older browsers.

h-m-jones commented on grails-app/taglib/uk/org/nbn/biocache/hubs/OccurrenceTagLib.groovy:
```diff
@@ -199,4 +199,99 @@ class OccurrenceTagLib extends au.org.ala.biocache.hubs.OccurrenceTagLib{
+        def targetSelector = attrs.targetSelector ?: 'body' // Default to body if no selector provided
+
+        // Output modal HTML directly to the page
+        out << g.render(template: '/occurrence/wmsModal')
```
just wondering why the javascript can't be in the template too - instead of writing to a String and outputting it like below?

tony-ross replied:
moved JS to template - much better

tony-ross committed:
Add WMS Button and Help & Feedback Panel Enhancements
- Introduced a new CSS file for WMS button styling.
- Updated OccurrenceTagLib to handle help panel toggle functionality.
- Integrated WMS button and help panel into the modal header in _wmsModal.gsp.
- Added help panel content for user feedback and feature explanation in the modal.

tony-ross committed:
Enhance WMS Button Modal with Tabs
- Added new styles for tab content and navigation tabs to improve layout.
- Introduced styling for geoJson textarea and content area for better user experience.

tony-ross committed:
- Refactored OccurrenceTagLib to move JS out to gsp
- Added new JavaScript functions for better handling of WMS parameters and GeoJSON data loading.
